aten/src/ATen/core/dispatch/Dispatcher.cpp

```diff
@@ -86,69 +88,123 @@ OperatorHandle Dispatcher::findOrRegisterSchema_(FunctionSchema&& schema) {
   // we need a lock to avoid concurrent writes
   std::lock_guard<std::mutex> lock(mutex_);
 
   OperatorName op_name = schema.operator_name();
 
   auto op = findOrRegisterSchema_(std::move(schema));
 
-  ++op.operatorIterator_->refcount;
-  if (1 == op.operatorIterator_->refcount) {
+  ++op.operatorIterator_->def_count;
+  ++op.operatorIterator_->def_and_impl_count;
+  if (1 == op.operatorIterator_->def_count) {
     // note: call listeners *after* operator is added, i.e. dispatcher is already valid for new op
     listeners_->callOnOperatorRegistered(op);
   }
 
-  return std::make_pair(RegistrationHandleRAII([this, op, op_name] {
-    deregisterSchema_(op, op_name);
-  }), op);
+  return RegistrationHandleRAII([this, op, op_name] {
+    deregisterDef_(op, op_name);
+  });
 }
 
-void Dispatcher::deregisterSchema_(const OperatorHandle& op, const OperatorName& op_name) {
+void Dispatcher::deregisterDef_(const OperatorHandle& op, const OperatorName& op_name) {
   // we need a lock to avoid concurrent writes
   std::lock_guard<std::mutex> lock(mutex_);
 
   TORCH_INTERNAL_ASSERT(op.schema().operator_name() == op_name);
 
-  // reduce refcount and actually deregister if no references left
-  TORCH_INTERNAL_ASSERT(op.operatorIterator_->refcount > 0);
-  --op.operatorIterator_->refcount;
-  if (0 == op.operatorIterator_->refcount) {
-    op.operatorIterator_->op.prepareForDeregistration();
-
+  // reduce def_count and actually deregister if no references left
+  TORCH_INTERNAL_ASSERT(op.operatorIterator_->def_count > 0);
+  TORCH_INTERNAL_ASSERT(op.operatorIterator_->def_and_impl_count > 0);
+  --op.operatorIterator_->def_count;
+  --op.operatorIterator_->def_and_impl_count;
+  if (0 == op.operatorIterator_->def_count) {
     // note: call listeners *before* operator is removed, i.e. dispatcher is still valid for removed op
+    // TODO: check that listeners are not relying on prepareForDeregistration()
+    // invariant
     listeners_->callOnOperatorDeregistered(op);
+  }
+  if (0 == op.operatorIterator_->def_and_impl_count) {
+    // TODO: rename this to "assert deregistration invariants"
+    op.operatorIterator_->op.prepareForDeregistration();
+    operators_.erase(op.operatorIterator_);
+    operatorLookupTable_.write([&] (ska::flat_hash_map<OperatorName, OperatorHandle>& operatorLookupTable) {
+      operatorLookupTable.erase(op_name);
+    });
+  }
+}
+
+RegistrationHandleRAII Dispatcher::registerImpl(OperatorName op_name, c10::optional<DispatchKey> dispatch_key, KernelFunction kernel) {
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  auto op = findOrRegisterName_(op_name);
+
+  auto kernel_handle = op.operatorIterator_->op.registerKernel(dispatch_key, std::move(kernel));
 
+  ++op.operatorIterator_->def_and_impl_count;
+
+  return RegistrationHandleRAII([this, op, op_name, dispatch_key, kernel_handle] {
+    op.operatorIterator_->op.deregisterKernel_(dispatch_key, kernel_handle);
```

Comment:
why doesn't `deregisterKernel_` get called from inside` deregisterImpl_` instead? Would be more consistent since it's `registerImpl` which calls `registerKernel` (and you could get rid of the comment in line 171+172)

Reply:
The reason is actually really dumb: `kernel_handle` is a big type (`std::list<KernelFunction>::iterator` in this PR) and if I put it in deregisterKernel then I have to add it to a bunch of extra type signatures. So I've been torn about just inlining all the deregister methods directly into the lambdas so I can reduce the amount of C++ boilerplate I have to write here.

Reply:
I still feel consistency here is more important. Boilerplate can be skipped when reading it but if things behave differently than you'd expect, that can cause confusion.